File: c++/max-area-of-island.cc
class Solution {
private:
  int scourIsland(vector<vector<int>>& grid, int i, int j) {
    int area = 0;
    queue<pair<int, int>> bfsq;
    bfsq.push({i, j});
    while (!bfsq.empty()) {
      auto coord = bfsq.front();
      bfsq.pop();
      if (!grid[coord.first][coord.second]) continue;

      area++;
      grid[coord.first][coord.second] = 0;

      if (coord.first > 0 && grid[coord.first - 1][coord.second]) {
        bfsq.push({coord.first - 1, coord.second});
      }
      if (coord.first < grid.size() - 1 && grid[coord.first + 1][coord.second]) {
        bfsq.push({coord.first + 1, coord.second});
      }
      if (coord.second  > 0 && grid[coord.first][coord.second  - 1]) {
        bfsq.push({coord.first, coord.second  - 1});
      }
      if (coord.second  < grid[0].size() - 1 && grid[coord.first][coord.second  + 1]) {
        bfsq.push({coord.first, coord.second  + 1});
      }
    }

    return area;
  }
public:
  int maxAreaOfIsland(vector<vector<int>>& grid) {
    int maxArea = 0;

    for (int i = 0; i < grid.size(); ++i) {
      for (int j = 0; j < grid[0].size(); ++j) {
        if (!grid[i][j]) continue;
        maxArea = max(maxArea, scourIsland(grid, i, j));
      }
    }

    return maxArea;  
  }
}

File: python/max-area-of-island.py
class Solution:
    def maxAreaOfIsland(self, grid: List[List[int]]) -> int:
        def exploreIsland(x: int, y: int) -> None:
            nonlocal curArea
            if y < 0 or x < 0 or y >= len(grid) or x >= len(grid[0]) or grid[y][x] == 0:
                return
            grid[y][x] = 0
            curArea += 1
            exploreIsland(x + 1, y)
            exploreIsland(x - 1, y)
            exploreIsland(x, y + 1)
            exploreIsland(x, y - 1)
        
        maxArea = 0
        for i in range(len(grid)):
            for j in range(len(grid[0])):
                if grid[i][j] == 1:
                    curArea = 0
                    exploreIsland(j, i)
                    maxArea = max(maxArea, curArea)

        
        return maxArea
